runtime/near-vm-logic/src/receipt_manager.rs

```diff
@@ -0,0 +1,493 @@
+    pub fn append_action_delete_account(
+        &mut self,
+        receipt_index: u64,
+        beneficiary_id: AccountId,
+    ) -> ExtResult<()> {
+        self.append_action(
+            receipt_index,
+            Action::DeleteAccount(DeleteAccountAction { beneficiary_id }),
+        );
+        Ok(())
+    }
+
+    /// Distribute the gas among the scheduled function calls that specify a gas weight.
+    ///
+    /// Distributes the gas passed in by splitting it among weights defined in `gas_weights`.
+    /// This will sum all weights, retrieve the gas per weight, then update each function
+    /// to add the respective amount of gas. Once all gas is distributed, the remainder of
+    /// the gas not assigned due to precision loss is added to the last function with a weight.
+    ///
+    /// # Arguments
+    ///
+    /// * `gas` - amount of unused gas to distribute
+    ///
+    /// # Returns
+    ///
+    /// Function returns a [GasDistribution] that indicates how the gas was distributed.
+    #[cfg(feature = "protocol_feature_function_call_weight")]
+    pub fn distribute_unused_gas(&mut self, gas: u64) -> GasDistribution {
+        let gas_weight_sum: u128 =
+            self.gas_weights.iter().map(|(_, GasWeight(weight))| *weight as u128).sum();
+        if gas_weight_sum != 0 {
+            // Floor division that will ensure gas allocated is <= gas to distribute
+            let gas_per_weight = (gas as u128 / gas_weight_sum) as u64;
+
+            let mut distribute_gas = |metadata: &FunctionCallActionIndex, assigned_gas: u64| {
+                let FunctionCallActionIndex { receipt_index, action_index } = metadata;
+                if let Some(Action::FunctionCall(FunctionCallAction { ref mut gas, .. })) = self
+                    .action_receipts
+                    .0
+                    .get_mut(*receipt_index)
+                    .and_then(|(_, receipt)| receipt.actions.get_mut(*action_index))
+                {
+                    *gas += assigned_gas;
+                } else {
+                    panic!(
+                        "Invalid index for assigning unused gas weight \
+                        (promise_index={}, action_index={})",
+                        receipt_index, action_index
+                    );
+                }
+            };
+
+            let mut distributed = 0;
+            for (action_index, GasWeight(weight)) in &self.gas_weights {
+                // This can't overflow because the gas_per_weight is floor division
+                // of the weight sum.
+                let assigned_gas = gas_per_weight * weight;
+
+                distribute_gas(action_index, assigned_gas);
+
+                distributed += assigned_gas
+            }
+
+            // Distribute remaining gas to final action.
+            if let Some((last_idx, _)) = self.gas_weights.last() {
+                distribute_gas(last_idx, gas - distributed);
+            }
```

Comment:
:thinking: I am still not happy about this algorithm. A bad example here would be if we want to distribute `10` gas across `10` function calls evenly, but where each weight is `1000`. In that situation, `gas_per_weight` will be zero, and only the last call will get the gas. 

I suggest tackling this separately though, let's first land this much needed refactor.

Reply:
So with this, this is more of an edge case where the weight is greater or close to the amount of gas remaining.

The downside to this is just that it requires extra logic and computation to handle for this edge case. I'm also wondering in the case of splitting 12 gas among the 10 function calls, do the last two functions receive an extra 1 gas or how will this happen?

I don't have an opinion on which direction this goes, I was just defaulting to this for simplicity rather than creating a bunch of semantics that only does something subjectively meaningful with improper usage.

Reply:
I don't think handlig this particular cases requires that much excar logic. I think what we need here is

```
let assigned_gas = (gas_to_distribute as u128 * weight as u128) / sum_weights;
```

That is, we just needlessly loosing precision here. 

Distributing the remainder evenly is a different, smaller problem, that I think would require more complex logic.

Reply:
This is what I'm referring to, the remainder distribution being more calculations.

My intuition is that it would divide the amount of remaining gas by the number of individual weights to get the distribution of gas per function call, then get the remainder of that division (`x`) to attach the remaining gas to final `x` number of function calls.

Seems like unnecessary logic and computation for what will be trivial amounts of gas unless specifically misused by specifying gas weights larger than the amount of gas, but I'm happy to make the change if there is consensus on this :)

Reply:
From what I see, here we choose between

```
1) let assigned_gas = ((gas_to_distribute as u128 * weight as u128) / sum_weights) as u64;
2) let assigned_gas = (gas_to_distribute / sum_weights) * weight;
```

I don't see a significant difference for regular usecases, so I would also choose (2) because it allows to make less operations.